modules/contracts/apps/subscription/src/handlers/execute.rs

```diff
@@ -0,0 +1,351 @@
+                * config.subscription_cost_per_week,
+        )?;
+        subscription_state.active_subs += 1;
+        SUBSCRIPTION_STATE.save(deps.storage, &subscription_state)?;
+    }
+
+    Ok(app.tag_response(
+        Response::new().add_attributes(attrs).add_message(
+            // Send the received asset to the proxy
+            asset.transfer_msg(base_state.proxy_address)?,
+        ),
+        "pay",
+    ))
+}
+
+pub fn unsubscribe(
+    deps: DepsMut,
+    env: Env,
+    app: SubscriptionApp,
+    unsubscribe_addrs: Vec<String>,
+) -> SubscriptionResult {
+    let unsubscribe_addrs: Vec<Addr> = unsubscribe_addrs
+        .iter()
+        .map(|human| deps.api.addr_validate(human))
+        .collect::<StdResult<_>>()?;
+    let mut subscription_state = SUBSCRIPTION_STATE.load(deps.storage)?;
+    let subscription_config = SUBSCRIPTION_CONFIG.load(deps.storage)?;
+    let mut unsubscribed_addrs: Vec<String> = vec![];
+    let mut claim_msgs: Vec<ExecutorMsg> = vec![];
+
+    // update income
+    INCOME_TWA.accumulate(
+        &env,
+        deps.storage,
+        Decimal::from_atomics(Uint128::from(subscription_state.active_subs), 0)?
+            * subscription_config.subscription_cost_per_week,
+    )?;
+
+    for addr in unsubscribe_addrs.iter() {
+        let mut subscriber = SUBSCRIBERS.load(deps.storage, addr)?;
+        if subscriber.expiration_timestamp <= env.block.time {
+            let maybe_claim_msg = match claim_emissions_msg(
+                &app,
+                deps.as_ref(),
+                &env,
+                &mut subscriber,
+                addr,
+                subscription_config.subscription_per_week_emissions.clone(),
+                &subscription_state,
+            ) {
+                Ok(maybe_msg) => maybe_msg,
+                // If just claimed or not enabled - no claims
+                Err(SubscriptionError::EmissionsAlreadyClaimed {})
+                | Err(SubscriptionError::SubscriberEmissionsNotEnabled {}) => None,
+                Err(error) => {
+                    return Err(error);
+                }
+            };
+
+            unsubscribed_addrs.push(addr.to_string());
+            subscription_state.active_subs -= 1;
+            SUBSCRIBERS.remove(deps.storage, addr);
+            EXPIRED_SUBSCRIBERS.save(deps.storage, addr, &subscriber)?;
+
+            if let Some(msg) = maybe_claim_msg {
+                claim_msgs.push(msg)
+            }
+        }
+    }
+
+    // Error if no one unsubbed
+    if unsubscribed_addrs.is_empty() {
+        return Err(SubscriptionError::NoOneUnsubbed {});
+    }
+
+    // update subscription count
+    SUBSCRIPTION_STATE.save(deps.storage, &subscription_state)?;
+
+    let mut response = app.tag_response(Response::new().add_messages(claim_msgs), "unsubscribe");
+
+    if let Some(hook) = subscription_config.unsubscription_hook_addr {
+        let msg = UnsubscribedHookMsg {
+            unsubscribed: unsubscribed_addrs.clone(),
```

Comment:
Do we need to clone here? I might be missing something but I don't see it used after this line :)

Reply:
Yep, it was vector of hookmsgs before, clippy didn't notice that, LMAO
@AndrewPochapsky >>>> Clippy